src/dsa/rpo_stark/stark/prover.rs

```diff
@@ -0,0 +1,145 @@
+    StarkDomain, Trace, TraceInfo, TracePolyTable, TraceTable,
+};
+use winter_utils::{Deserializable, Serializable};
+use winterfell::{
+    AuxRandElements, CompositionPoly, CompositionPolyTrace, ConstraintCompositionCoefficients,
+    DefaultConstraintCommitment, PartitionOptions,
+};
+
+use super::air::{PublicInputs, RescueAir, HASH_CYCLE_LEN};
+use crate::{hash::rpo::Rpo256, rand::RpoRandomCoin, Word, ZERO};
+
+// PROVER
+// ================================================================================================
+
+pub struct RpoSignatureProver<H: ElementHasher>
+where
+    H: Sync,
+{
+    options: ProofOptions,
+    _hasher: PhantomData<H>,
+}
+
+impl<H: ElementHasher + Sync> RpoSignatureProver<H> {
+    pub fn new(options: ProofOptions) -> Self {
+        Self { options, _hasher: PhantomData }
+    }
+
+    pub fn build_trace(&self, sk: Word, msg: Word) -> TraceTable<BaseElement> {
+        let trace_length = HASH_CYCLE_LEN;
+        let mut target = vec![];
+        msg.write_into(&mut target);
+        let mut trace = TraceTable::with_meta(12, trace_length, target);
+
+        trace.fill(
+            |state| {
+                // initialize first state of the computation
+                state[0] = ZERO;
+                state[1] = ZERO;
+                state[2] = ZERO;
+                state[3] = ZERO;
+                state[4] = sk[0];
+                state[5] = sk[1];
+                state[6] = sk[2];
+                state[7] = sk[3];
+                state[8] = ZERO;
+                state[9] = ZERO;
+                state[10] = ZERO;
+                state[11] = ZERO;
+            },
+            |step, state| {
+                Rpo256::apply_round(
+                    state.try_into().expect("should not fail given the size of the array"),
+                    step,
+                );
+            },
+        );
+        trace
+    }
+}
+
+impl<H: ElementHasher> Prover for RpoSignatureProver<H>
+where
+    H: ElementHasher<BaseField = BaseElement> + Sync,
+{
+    type BaseField = BaseElement;
+    type Air = RescueAir;
+    type Trace = TraceTable<BaseElement>;
+    type HashFn = Rpo256;
+    type VC = SaltedMerkleTree<Self::HashFn, Self::ZkPrng>;
+    type RandomCoin = RpoRandomCoin;
+    type TraceLde<E: FieldElement<BaseField = Self::BaseField>> =
+        DefaultTraceLde<E, Self::HashFn, Self::VC>;
+    type ConstraintCommitment<E: FieldElement<BaseField = Self::BaseField>> =
+        DefaultConstraintCommitment<E, Self::HashFn, Self::ZkPrng, Self::VC>;
+    type ConstraintEvaluator<'a, E: FieldElement<BaseField = Self::BaseField>> =
+        DefaultConstraintEvaluator<'a, Self::Air, E>;
+    type ZkPrng = ChaCha20Rng;
+
+    fn get_pub_inputs(&self, trace: &Self::Trace) -> PublicInputs {
+        let last_step = trace.length() - 1;
+        let source = trace.info().meta();
+        let msg = <Word>::read_from_bytes(source).expect("the message should be a Word");
```

Comment:
One thing that took me a while to figure out (and I'm not sure I'm 100% correct), is that we don't encode the message into the trace itself. Instead, we put it into the public inputs, which influences the transcript and therefore the randomness used during proof generation. If this is correct, I would probably add an explicit comment here (and maybe in some other places), to explain it.

Reply:
Your understanding is spot on, this is a subtle point in that the message is part of the public inputs but it is not a part of the execution trace. This is something that is not supported by Winterfell and hence had to resort to this hack.  One of the issues with this solution is that the meta data is treated as a `&[u8]` but this makes it difficult to serialize and de-serialize the word representing the message.
Curious if you have a cleaner way to deal with this.

Reply:
Great idea, implemented!
I also added a comment to clarify what is happening here